Title: immudb 1.11.0 SQL engine panic (nil pointer dereference / index out of range) on valid queries

**What happened**
After upgrading from immudb 1.10.0 to 1.11.0, valid SQL queries started causing immudb to panic.
Downgrading back to 1.10.0 immidiately resolves the issue without any code or data changes.

Panic 1:
Panic: runtime error: invalid memory address or nil pointer dereference

Stack trace:
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0xb4c540]
goroutine 80357 [running]:
github.com/codenotary/immudb/embedded/sql.encodeRow(0xc0511bf460)
/src/embedded/sql/file_sort.go:455 +0xa0
github.com/codenotary/immudb/embedded/sql.(*fileSorter).flushBuffer(0xc035569c40)
/src/embedded/sql/file_sort.go:415 +0xa8
github.com/codenotary/immudb/embedded/sql.(*fileSorter).sortAndFlushBuffer(0xc035569c40)
/src/embedded/sql/file_sort.go:387 +0x2d
github.com/codenotary/immudb/embedded/sql.(*fileSorter).update(0xc035569c40, 0xc032da41c0)
/src/embedded/sql/file_sort.go:60 +0x2b
github.com/codenotary/immudb/embedded/sql.(*sortRowReader).readAll(0xc035569c00, {0x2ba91c8, 0xc03a79fda0})
/src/embedded/sql/sort_reader.go:397 +0x99
github.com/codenotary/immudb/embedded/sql.(*sortRowReader).readAndSort(0xc035569c00, {0x2ba91c8?, 0xc03a79fda0?})
/src/embedded/sql/sort_reader.go:298 +0x2c
github.com/codenotary/immudb/embedded/sql.(*sortRowReader).Read(0xc035569c00, {0x2ba91c8?, 0xc03a79fda0?})
/src/embedded/sql/sort_reader.go:285 +0x30
github.com/codenotary/immudb/embedded/sql.(*projectedRowReader).Read(0xc03ec3cf00, {0x2ba91c8?, 0xc03a79fda0?})
/src/embedded/sql/proj_row_reader.go:186 +0x42
github.com/codenotary/immudb/embedded/sql.(*offsetRowReader).Read(0xc0511bf1c0, {0x2ba91c8, 0xc03a79fda0})
/src/embedded/sql/offset_row_reader.go:73 +0x42
github.com/codenotary/immudb/embedded/sql.(*limitRowReader).Read(0xc0511bf1e0, {0x2ba91c8?, 0xc03a79fda0?})
/src/embedded/sql/limit_row_reader.go:76 +0x3b
github.com/codenotary/immudb/pkg/database.(*limitRowReader).Read(0xc0511bf200, {0x2ba91c8?, 0xc03a79fda0?})
/src/pkg/database/sql.go:482 +0x2e
github.com/codenotary/immudb/embedded/sql.readNRows({0x2ba91c8, 0xc03a79fda0}, {0x2bb5a20, 0xc0511bf200}, 0x9c4, {0xc0310af008, 0x9c4, 0x41d3e5?})
/src/embedded/sql/row_reader.go:684 +0x7c
github.com/codenotary/immudb/embedded/sql.ReadRowsBatch({0x2ba91c8, 0xc03a79fda0}, {0x2bb5a20, 0xc0511bf200}, 0x9c4, 0xc01ba2d1e0)
/src/embedded/sql/row_reader.go:666 +0x85
github.com/codenotary/immudb/pkg/server.(*ImmuServer).streamRows(0xc03106dc30?, {0x2ba91c8, 0xc03a79fda0}, {0x2bb5a20, 0xc0511bf200}, 0x9c4, 0xc01ba2d3f0)
/src/pkg/server/sql.go:161 +0x3c9
github.com/codenotary/immudb/pkg/server.(*ImmuServer).sqlQuery(0xc0000f4160, {0x2ba91c8, 0xc03a79fda0}, 0xc04c816960, 0xc01ba2d3f0)
/src/pkg/server/sql.go:143 +0x23d
github.com/codenotary/immudb/pkg/server.(*ImmuServer).SQLQuery(0xc0000f4160, 0xc04c816960, {0x2bb1750, 0xc03106dbb0})
/src/pkg/server/sql.go:147 +0x67
github.com/codenotary/immudb/pkg/api/schema._ImmuService_SQLQuery_Handler({0x1389420, 0xc0000f4160}, {0x2baf150, 0xc053ac4330})
/src/pkg/api/schema/schema_grpc.pb.go:2757 +0x113
github.com/codenotary/immudb/pkg/server.(*ImmuServer).InjectRequestMetadataStreamInterceptor(0xc0000f4160, {0x1389420, 0xc0000f4160}, {0x2baf150, 0xc053ac4330}, 0xc03106db50?, 0x2a1e520)
/src/pkg/server/request_metadata_interceptor.go:23 +0x102
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainStreamServer.func6.1.1({0x1389420?, 0xc0000f4160?}, {0x2baf150?, 0xc053ac4330?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:49 +0x34
github.com/codenotary/immudb/pkg/auth.ServerStreamInterceptor({0x1389420, 0xc0000f4160}, {0x2baf150, 0xc053ac4330}, 0x12dcfe0?, 0xc0511be620)
/src/pkg/auth/serverinterceptors.go:51 +0xa2
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainStreamServer.func6.1.1({0x1389420?, 0xc0000f4160?}, {0x2baf150?, 0xc053ac4330?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:49 +0x34
github.com/grpc-ecosystem/go-grpc-prometheus.init.(*ServerMetrics).StreamServerInterceptor.func4({0x1389420, 0xc0000f4160}, {0x2baeb20, 0xc035569900}, 0xc053ac4318, 0xc0511be640)
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:121 +0xd2
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainStreamServer.func6.1.1({0x1389420?, 0xc0000f4160?}, {0x2baeb20?, 0xc035569900?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:49 +0x34
github.com/codenotary/immudb/pkg/server.(*uuidContext).UUIDStreamContextSetter(0x47e3c5?, {0x1389420, 0xc0000f4160}, {0x2baeb20, 0xc035569900}, 0xc030cb1720?, 0xc0511be660)
/src/pkg/server/uuid.go:125 +0x126
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainStreamServer.func6.1.1({0x1389420?, 0xc0000f4160?}, {0x2baeb20?, 0xc035569900?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:49 +0x34
github.com/codenotary/immudb/pkg/server.(*ImmuServer).KeepALiveSessionStreamInterceptor(0xc0000f4160, {0x1389420, 0xc0000f4160}, {0x2baeb20, 0xc035569900}, 0x37a57c0?, 0xc0511be680)
/src/pkg/server/keep_alive_session_interceptor.go:33 +0x8b
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainStreamServer.func6.1.1({0x1389420?, 0xc0000f4160?}, {0x2baeb20?, 0xc035569900?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:49 +0x34
github.com/codenotary/immudb/pkg/server.(*ImmuServer).AccessLogStreamInterceptor(0xc0000f4160, {0x1389420, 0xc0000f4160}, {0x2baeb20, 0xc035569900}, 0xc053ac4318, 0xc0511be6a0)
/src/pkg/server/access_log_interceptor.go:13 +0x8e
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainStreamServer.func6.1.1({0x1389420?, 0xc0000f4160?}, {0x2baeb20?, 0xc035569900?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:49 +0x34
github.com/codenotary/immudb/pkg/server.(*ImmuServer).AuditLogStreamInterceptor(0x40f627?, {0x1389420?, 0xc0000f4160?}, {0x2baeb20?, 0xc035569900?}, 0x0?, 0x0?)
/src/pkg/server/audit_log_interceptor.go:43 +0x177
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainStreamServer.func6.1.1({0x1389420?, 0xc0000f4160?}, {0x2baeb20?, 0xc035569900?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:49 +0x34
github.com/codenotary/immudb/pkg/server.ErrorMapperStream({0x1389420?, 0xc0000f4160?}, {0x2baeb20?, 0xc035569900?}, 0xc01124f808?, 0xc053ac4318?)
/src/pkg/server/error_mapper_interceptor.go:27 +0x2a
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainStreamServer.func6.1.1({0x1389420?, 0xc0000f4160?}, {0x2baeb20?, 0xc035569900?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:49 +0x34
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainStreamServer.func6({0x1389420, 0xc0000f4160}, {0x2baeb20, 0xc035569900}, 0x11349a0?, 0xc03ec3c9c0?)
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:58 +0xaa
google.golang.org/grpc.(*Server).processStreamingRPC(0xc000540b48, {0x2ba91c8, 0xc03a79fd40}, 0xc02ec3c820, 0xc00dd2b470, 0x37606e0, 0x0)
/go/pkg/mod/google.golang.org/grpc@v1.79.3/server.go:1722 +0x1151
google.golang.org/grpc.(*Server).handleStream(0xc000540b48, {0x2ba9f58, 0xc000322680}, 0xc02ec3c820)
/go/pkg/mod/google.golang.org/grpc@v1.79.3/server.go:1859 +0xb85
google.golang.org/grpc.(*Server).serveStreams.func2.1()
/go/pkg/mod/google.golang.org/grpc@v1.79.3/server.go:1064 +0x7f
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 80133
/go/pkg/mod/google.golang.org/grpc@v1.79.3/server.go:1075 +0x11d

SQL Query Example: SELECT column1,column2,... from table1 where country=@country ORDER BY id LIMIT @set_limit OFFSET @set_offset

**Reproduction Behaviour**

Dataset characteristics:
- Table contains approximately 40K+ records.
- Country coulmn is indexed.
- Data was loaded in batches over time.
- A large batch (~25K records) was inserted recently.

Observed behavior on immudb 1.11.0:
works:
- Querying all users (no country filter).
- Querying countries whose records were inserted earlier in the table/database.
- Querying a combination of countries that includes both:
   An earlier inserted country , and 
   A later inserted country records.

Fails:
- Querying a single country whose records were inserted later in the database.
- Querying a countries that mainly contain recently records.

Validation Performed:
- CPU and memory utilization are low during failure.
- same dataset imported into local immudb 1.10.0 works
- AWS immudb 1.11.0 reproduces.
- Downgrading AWS immudb to 1.10.0 resolves all failures.
- Queries themselves are valid and return expected results on 1.10.0

Possible reproduction pattern:
The issue apprears related to filtering over records that were inserted later in the dataset and may depend on how immudb processes indexed lookups,sorting, or result set generation for those rows.
The Same logical query behaves differently depending on whether it targets older records or more recently inserted records.

**What you expected to happen**
Queries should either:
- return results, or 
- return a normal SQL/database error.
- They should not panic the immudb server.



Additional regression in immudb 1.11.0
I found a second SQL engine panic after upgrading from immudb 1.10.0 to 1.11.0.

Stack trace:
github.com/codenotary/immudb/pkg/api/schema._ImmuService_UnarySQLQuery_Handler({0x1389420, 0xc0000f4160}, {0x2ba91c8, 0xc06a500810}, 0xc05f650480, 0xc00dd6a5d0)
/src/pkg/api/schema/schema_grpc.pb.go:2749 +0x143
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00024cb48, {0x2ba91c8, 0xc06a500780}, 0xc0562f2680, 0xc00dd6a8d0, 0x3766140, 0x0)
/go/pkg/mod/google.golang.org/grpc@v1.79.3/server.go:1429 +0x1192
google.golang.org/grpc.(*Server).handleStream(0xc00024cb48, {0x2ba9f58, 0xc0402c0000}, 0xc0562f2680)
/go/pkg/mod/google.golang.org/grpc@v1.79.3/server.go:1855 +0xbc6
google.golang.org/grpc.(*Server).serveStreams.func2.1()
/go/pkg/mod/google.golang.org/grpc@v1.79.3/server.go:1064 +0x7f
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 553255
/go/pkg/mod/google.golang.org/grpc@v1.79.3/server.go:1075 +0x11d
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.QueryLatencyInterceptor.func3({0x2ba91c8, 0xc06a500810}, {0x12e2900, 0xc059f48420}, 0xc05f64e7c0, 0xc05f64e820)
/src/pkg/server/metrics.go:397 +0x82
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainUnaryServer.func4.1.1({0x2ba91c8?, 0xc06a500810?}, {0x12e2900?, 0xc059f48420?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x34
github.com/grpc-ecosystem/go-grpc-prometheus.init.(*ServerMetrics).UnaryServerInterceptor.func3({0x2ba91c8, 0xc06a500810}, {0x12e2900, 0xc059f48420}, 0xc05f630b58?, 0xc05f64e840)
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:107 +0x77
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainUnaryServer.func4.1.1({0x2ba91c8?, 0xc06a500810?}, {0x12e2900?, 0xc059f48420?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x34
github.com/codenotary/immudb/pkg/server.(*uuidContext).UUIDContextSetter(0x4?, {0x2ba91c8, 0xc06a500810}, {0x12e2900, 0xc059f48420}, 0x139bf23?, 0xc05f64e860)
/src/pkg/server/uuid.go:137 +0x12f
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainUnaryServer.func4.1.1({0x2ba91c8?, 0xc06a500810?}, {0x12e2900?, 0xc059f48420?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x34
github.com/codenotary/immudb/pkg/server.(*ImmuServer).KeepAliveSessionInterceptor(0xc0000f4160, {0x2ba91c8, 0xc06a500810}, {0x12e2900, 0xc059f48420}, 0xc05f64e7c0, 0xc05f64e880)
/src/pkg/server/keep_alive_session_interceptor.go:44 +0xc3
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainUnaryServer.func4.1.1({0x2ba91c8?, 0xc06a500810?}, {0x12e2900?, 0xc059f48420?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x34
github.com/codenotary/immudb/pkg/server.(*ImmuServer).AccessLogInterceptor(0xc0000f4160, {0x2ba91c8, 0xc06a500810}, {0x12e2900, 0xc059f48420}, 0xc05f64e7c0, 0xc05f64e8a0)
/src/pkg/server/access_log_interceptor.go:22 +0x8e
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainUnaryServer.func4.1.1({0x2ba91c8?, 0xc06a500810?}, {0x12e2900?, 0xc059f48420?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x34
github.com/codenotary/immudb/pkg/server.(*ImmuServer).AuditLogInterceptor(0xc059f48420?, {0x2ba91c8?, 0xc06a500810?}, {0x12e2900?, 0xc059f48420?}, 0x7f0d66df2108?, 0x8?)
/src/pkg/server/audit_log_interceptor.go:31 +0x174
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainUnaryServer.func4.1.1({0x2ba91c8?, 0xc06a500810?}, {0x12e2900?, 0xc059f48420?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x34
github.com/codenotary/immudb/pkg/server.ErrorMapper({0x2ba91c8?, 0xc06a500810?}, {0x12e2900?, 0xc059f48420?}, 0xc00006e808?, 0xc05f64e7c0?)
/src/pkg/server/error_mapper_interceptor.go:33 +0x2a
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainUnaryServer.func4.1.1({0x2ba91c8?, 0xc06a500810?}, {0x12e2900?, 0xc059f48420?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x34
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainUnaryServer.func4({0x2ba91c8, 0xc06a500810}, {0x12e2900, 0xc059f48420}, 0x41d3e5?, 0x78?)
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:34 +0xaa
panic: runtime error: index out of range [0] with length 0
goroutine 570897 [running]:
github.com/codenotary/immudb/pkg/database.(*db).SQLQuery(0xc000372100, {0x2ba91c8, 0xc06a500810}, 0xc05f549a70, 0xc059f48420)
/src/pkg/database/sql.go:403 +0x3af
github.com/codenotary/immudb/pkg/database.(*lazyDB).SQLQuery(0xc0369a9140, {0x2ba91c8, 0xc06a500810}, 0xc05f549a70, 0xc059f48420)
/src/pkg/database/lazy_db.go:354 +0xea
github.com/codenotary/immudb/pkg/server.(*ImmuServer).sqlQuery(0xc0000f4160, {0x2ba91c8, 0xc06a500810}, 0xc059f48420, 0xc01b1ed080)
/src/pkg/server/sql.go:133 +0x18d
github.com/codenotary/immudb/pkg/server.(*ImmuServer).UnarySQLQuery(0x373e560?, {0x2ba91c8?, 0xc06a500810?}, 0x10?)
/src/pkg/server/sql.go:114 +0x45
github.com/codenotary/immudb/pkg/api/schema._ImmuService_UnarySQLQuery_Handler.func1({0x2ba91c8?, 0xc06a500810?}, {0x12e2900?, 0xc059f48420?})
/src/pkg/api/schema/schema_grpc.pb.go:2747 +0xce
github.com/codenotary/immudb/pkg/server.(*ImmuServer).InjectRequestMetadataUnaryInterceptor(0x20?, {0x2ba91c8?, 0xc06a500810?}, {0x12e2900?, 0xc059f48420?}, 0x139bf23?, 0xd?)
/src/pkg/server/request_metadata_interceptor.go:15 +0x74
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainUnaryServer.func4.1.1({0x2ba91c8?, 0xc06a500810?}, {0x12e2900?, 0xc059f48420?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x34
github.com/codenotary/immudb/pkg/server.(*ImmuServer).SessionAuthInterceptor(0xc0000f4160, {0x2ba91c8, 0xc06a500810}, {0x12e2900, 0xc059f48420}, 0xc05f64e7c0, 0xc05f64e7e0)
/src/pkg/server/session_auth_interceptor.go:37 +0xf8
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainUnaryServer.func4.1.1({0x2ba91c8?, 0xc06a500810?}, {0x12e2900?, 0xc059f48420?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x34
github.com/codenotary/immudb/pkg/auth.ServerUnaryInterceptor({0x2ba91c8, 0xc06a500810}, {0x12e2900, 0xc059f48420}, 0xc00e4592b8?, 0xc05f64e800)
/src/pkg/auth/serverinterceptors.go:72 +0x9a
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Initialize.ChainUnaryServer.func4.1.1({0x2ba91c8?, 0xc06a500810?}, {0x12e2900?, 0xc059f48420?})
/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x34

SQL Query Example: Multiple left JOINs against the same table

**Observed Behaviour**
immjudb 1.10.0
- Query executes successfully

immjudb 1.11.0
- Server panic with: runtime error: index out of range [0] with length 0

**Environment**

- immudb 1.11.0
- AWS ESC Fargate
- Next.js/Node.js application
- Database size: ~40K+ users
- Country Coulumn Indexed

--- a/embedded/sql/file_sort.go
+++ b/embedded/sql/file_sort.go
@@ -412,7 +412,7 @@ func (s *fileSorter) flushBuffer() error {
 
 	var chunkSize uint64
 	for _, row := range s.sortBuf[:s.nextIdx] {
-		data, err := encodeRow(row)
+		data, err := s.encodeRow(row)
 		if err != nil {
 			return err
 		}
@@ -447,11 +447,18 @@ func (s *fileSorter) tempFileWriter() (*bufio.Writer, error) {
 	return s.writer, nil
 }
 
-func encodeRow(r *Row) ([]byte, error) {
+func (s *fileSorter) encodeRow(r *Row) ([]byte, error) {
 	var buf bytes.Buffer
 	buf.Write([]byte{0, 0}) // make room for size field
 
-	for _, v := range r.ValuesByPosition {
+	for i, v := range r.ValuesByPosition {
+		// Projection pushdown (rawRowReader.Read) leaves nil slots for table
+		// columns the query does not need; the nil zero-value is semantically
+		// NULL. Substitute a typed NULL so encoding never dereferences nil and
+		// the slot round-trips through the temp file as NULL.
+		if v == nil {
+			v = &NullValue{t: s.colTypes[i]}
+		}
 		rawValue, err := EncodeNullableValue(v, v.Type(), -1)
 		if err != nil {
 			return nil, err

--- a/embedded/sql/file_sort_pushdown_regression_test.go
+++ b/embedded/sql/file_sort_pushdown_regression_test.go
@@ -1,0 +1,96 @@
+/*
+Copyright 2026 Codenotary Inc. All rights reserved.
+
+SPDX-License-Identifier: BUSL-1.1
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://mariadb.com/bsl11/
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sql
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+// TestSortSpillWithProjectionPushdownDoesNotPanic reproduces issue #2100
+// (panic #1): a SELECT that does NOT reference every table column, over a
+// result set large enough to spill the sort buffer to a temp file, used to
+// panic with a nil pointer dereference in encodeRow.
+//
+// Projection pushdown (rawRowReader.Read) leaves row.ValuesByPosition[i] == nil
+// for columns excluded by neededColIDs. The disk-spill sort path
+// (fileSorter.flushBuffer -> encodeRow) iterates every position and called
+// v.Type() on those nil slots, crashing the whole server.
+//
+// The reproduction query shape from the issue is:
+//
+//	SELECT a, b FROM t WHERE ... ORDER BY id LIMIT n OFFSET m
+//
+// The OFFSET clause disables the top-N heap optimisation (it requires
+// offset == nil), forcing the full file-sort path; a small sortBufferSize plus
+// enough rows forces the spill where the panic lived.
+func TestSortSpillWithProjectionPushdownDoesNotPanic(t *testing.T) {
+	engine := setupCommonTest(t)
+	engine.sortBufferSize = 8
+
+	ctx := context.Background()
+	mustExec := func(stmt string) {
+		t.Helper()
+		_, _, err := engine.Exec(ctx, nil, stmt, nil)
+		require.NoError(t, err, stmt)
+	}
+
+	// "unused" is NOT referenced by the query below, so projection pushdown
+	// excludes it from neededColIDs and leaves its position nil on every row.
+	mustExec(`CREATE TABLE events (
+		id INTEGER NOT NULL AUTO_INCREMENT,
+		grp INTEGER NOT NULL,
+		payload VARCHAR[256] NOT NULL,
+		unused VARCHAR[256] NOT NULL,
+		PRIMARY KEY(id)
+	);`)
+
+	const nRows = 50
+	for i := 1; i <= nRows; i++ {
+		mustExec(fmt.Sprintf(
+			`INSERT INTO events (grp, payload, unused) VALUES (%d, 'payload_%d', 'unused_%d');`,
+			i, i, i,
+		))
+	}
+
+	// SELECT references id, grp and payload but NOT "unused" => the unused
+	// column is nil in ValuesByPosition. OFFSET forces the full file-sort and
+	// the small buffer forces a disk spill.
+	rows, err := engine.queryAll(ctx, nil,
+		`SELECT id, payload FROM events ORDER BY grp ASC LIMIT 20 OFFSET 0;`,
+		nil,
+	)
+	require.NoError(t, err, "#2100: sort spill with projection pushdown must not panic/error")
+	require.Len(t, rows, 20)
+
+	for i, r := range rows {
+		idVal, ok := r.ValuesByPosition[0].RawValue().(int64)
+		require.Truef(t, ok, "row %d: id should be int64, got %T", i, r.ValuesByPosition[0].RawValue())
+		// grp == id in this dataset, and ORDER BY grp ASC LIMIT 20 returns the
+		// first 20 inserted rows in order.
+		require.Equalf(t, int64(i+1), idVal, "row %d: rows must be in ORDER BY grp order", i)
+
+		field := r.ValuesByPosition[1]
+		require.NotNilf(t, field, "row %d: payload must not be nil (id=%d)", i, idVal)
+		require.Falsef(t, field.IsNull(), "row %d: payload must not be NULL (id=%d)", i, idVal)
+		require.Equalf(t, fmt.Sprintf("payload_%d", idVal), field.RawValue().(string),
+			"row %d: payload must round-trip through sort spill for id=%d", i, idVal)
+	}
+}

--- a/pkg/database/sql.go
+++ b/pkg/database/sql.go
@@ -400,6 +400,10 @@ func (d *db) SQLQuery(ctx context.Context, tx *sql.SQLTx, req *schema.SQLQueryRe
 		return nil, err
 	}
 
+	if len(stmts) == 0 {
+		return nil, sql.ErrExpectingDQLStmt
+	}
+
 	stmt, ok := stmts[0].(sql.DataSource)
 	if !ok {
 		return nil, sql.ErrExpectingDQLStmt

--- a/pkg/database/sql_emptyquery_regression_test.go
+++ b/pkg/database/sql_emptyquery_regression_test.go
@@ -1,0 +1,44 @@
+/*
+Copyright 2026 Codenotary Inc. All rights reserved.
+
+SPDX-License-Identifier: BUSL-1.1
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://mariadb.com/bsl11/
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package database
+
+import (
+	"context"
+	"testing"
+
+	"github.com/codenotary/immudb/embedded/sql"
+	"github.com/codenotary/immudb/pkg/api/schema"
+	"github.com/stretchr/testify/require"
+)
+
+// TestSQLQueryEmptyStatementDoesNotPanic reproduces issue #2100 (panic #2):
+// db.SQLQuery indexed stmts[0] without a length guard. Since 1.11.0 the SQL
+// grammar accepts empty / comment-only input and ParseSQL returns an empty
+// statement slice with no error, so an empty query (routine probe traffic from
+// drivers) panicked with "index out of range [0] with length 0", crashing the
+// server. It must return ErrExpectingDQLStmt instead.
+func TestSQLQueryEmptyStatementDoesNotPanic(t *testing.T) {
+	db := makeDb(t)
+
+	for _, q := range []string{"", "   ", "\n\t ", "-- just a comment", "/* nothing here */"} {
+		_, err := db.SQLQuery(context.Background(), nil, &schema.SQLQueryRequest{Sql: q})
+		require.ErrorIs(t, err, sql.ErrExpectingDQLStmt, "empty query %q must not panic", q)
+
+		_, err = db.SQLQueryAll(context.Background(), nil, &schema.SQLQueryRequest{Sql: q})
+		require.ErrorIs(t, err, sql.ErrExpectingDQLStmt, "empty query %q must not panic", q)
+	}
+}

